src/psyclone/psyir/transformations/replace_induction_variables.py

```diff
@@ -0,0 +1,272 @@
+                                  Loop, Reference)
+from psyclone.psyir.transformations.transformation_error \
+    import TransformationError
+
+
+class ReplaceInductionVariables(Transformation):
+    '''Move all supported induction variables out of the loop, and replace
+    their usage inside the loop. For example:
+
+    >>> from psyclone.psyir.frontend.fortran import FortranReader
+    >>> from psyclone.psyir.nodes import Loop
+    >>> from psyclone.psyir.transformations import ReplaceInductionVariables
+    >>> from psyclone.psyir.backend.fortran import FortranWriter
+    >>> psyir = FortranReader().psyir_from_source("""
+    ... subroutine sub()
+    ...     integer :: i, im, ic, tmp(100)
+    ...     do i=1, 100
+    ...         im = i - 1
+    ...         ic = 2
+    ...         tmp(i) = ic * im
+    ...     enddo
+    ... end subroutine sub""")
+    >>> loop = psyir.walk(Loop)[0]
+    >>> ReplaceInductionVariables().apply(loop)
+    >>> print(FortranWriter()(psyir))
+
+    will generate:
+
+    .. code-block:: fortran
+
+        subroutine sub()
+          integer :: i
+          integer :: im
+          integer :: ic
+          integer, dimension(100) :: tmp
+
+          do i = 1, 100, 1
+            tmp(i) = 2 * (i - 1)
+          enddo
+          ic = 2
+          im = i - 1 - 1
+
+        end subroutine sub
+
+    After the loop the replaced assignments to the induction variables are
+    added so these variables will have the correct value if they should be
+    used elsewhere.
+
+    The following restrictions apply for the assignment to an induction
+    variable:
+
+    * the variable must be a non-structure scalar.
+    * none of variables on the right-hand side can be written in the loop body
+      (the loop variable is written in the Loop statement, not in
+      the body, so it can be used).
+    * the assigned variable must not be read before the assignment.
+    * the assigned variable cannot occur on the right-hand side
+      (e.g. `k = k + 3`).
+    * there must only be one assignment to the variable.
+
+    '''
+    def __str__(self):
+        return "Replaces all induction variables in a loop."
+
+    # ------------------------------------------------------------------------
+    @staticmethod
+    def _replace_references(psyir, original, replacement):
+        '''This function replaces all occurrences of the `original` reference
+        with a copy of the node of the `replacement` in the given PSyIR tree.
+
+        :param psyir: the PSyIR in which to replace `original` with \
+            `replacement`.
+        :type psyir: :py:class:`psyclone.psyir.nodes.Node`
+        :param original: the reference which is to be replaced.
+        :type original: :py:class:`psyclone.psyir.nodes.Reference`
+        :param replacement: the node to insert.
+        :type replacement: :py:class:`psyclone.psyir.nodes.Node`
+
+        '''
+        # Just in case we avoid modifying the tree while walking it,
```

Comment:
Add a comma after "Just in case".

And is this really a problem? I think it shouldn't be

Reply:
After a lot of thinking I agreed, and I combined the two loops. The only thing I could think of is using an array twice (`a(a(i))`) somewhere. But even this shouldn't be a problem - the outer one will be replaced, and if it then handles the inner one (which is not not part of the tree anymore, who cares :)